extension/httpfs/httpfs.cpp

```diff
@@ -30,23 +36,86 @@ HTTPParams HTTPParams::ReadFrom(FileOpener *opener) {
 		timeout = DEFAULT_TIMEOUT;
 	}
 
-	return {timeout};
+	if (opener->TryGetCurrentSetting("http_retries", value)) {
+		retries = value.GetValue<uint64_t>();
+	} else {
+		retries = DEFAULT_RETRIES;
+	}
+
+	if (opener->TryGetCurrentSetting("http_retry_wait_ms", value)) {
+		retry_wait_ms = value.GetValue<uint64_t>();
+	} else {
+		retry_wait_ms = DEFAULT_RETRY_WAIT_MS;
+	}
+
+	if (opener->TryGetCurrentSetting("http_retry_backoff", value)) {
+		retry_backoff = value.GetValue<float>();
+	} else {
+		retry_backoff = DEFAULT_RETRY_BACKOFF;
+	}
+
+	return {timeout, retries, retry_wait_ms, retry_backoff};
 }
 
 void HTTPFileSystem::ParseUrl(string &url, string &path_out, string &proto_host_port_out) {
 	if (url.rfind("http://", 0) != 0 && url.rfind("https://", 0) != 0) {
-		throw std::runtime_error("URL needs to start with http:// or https://");
+		throw IOException("URL needs to start with http:// or https://");
 	}
 	auto slash_pos = url.find('/', 8);
 	if (slash_pos == string::npos) {
-		throw std::runtime_error("URL needs to contain a '/' after the host");
+		throw IOException("URL needs to contain a '/' after the host");
 	}
 	proto_host_port_out = url.substr(0, slash_pos);
 
 	path_out = url.substr(slash_pos);
 
 	if (path_out.empty()) {
-		throw std::runtime_error("URL needs to contain a path");
+		throw IOException("URL needs to contain a path");
+	}
+}
+
+// Retry the request performed by fun using the exponential backoff strategy defined in params. Before retry, the
+// retry callback is called
+static unique_ptr<ResponseWrapper>
+RunRequestWithRetry(const std::function<duckdb_httplib_openssl::Result(void)> &request, string &url, string method,
+                    const HTTPParams &params, const std::function<void(void)> &retry_cb = {}) {
+	idx_t tries = 0;
+	while (true) {
+		auto res = request();
+
+		// Connection level errors to retry
+		bool should_retry = res.error() == duckdb_httplib_openssl::Error::Read;
+		should_retry = should_retry || res.error() == duckdb_httplib_openssl::Error::Connection;
+		should_retry = should_retry || res.error() == duckdb_httplib_openssl::Error::Write;
+
+		if (res.error() == duckdb_httplib_openssl::Error::Success) {
+			switch (res->status) {
+			case 408: // Request Timeout
+			case 418: // Server is pretending to be a teapot
+			case 429: // Rate limiter hit
+			case 503: // Server has error
+			case 504: // Server has error
+				should_retry = true;
+				break;
+			default:
+				return make_unique<ResponseWrapper>(res.value());
+			}
+		}
+
+		tries += 1;
+
+		if (should_retry && tries <= params.retries) {
+			if (tries > 1) {
+				uint64_t sleep_amount = (uint64_t)((float)params.retry_wait_ms * pow(params.retry_backoff, tries - 2));
+				std::this_thread::sleep_for(std::chrono::milliseconds(sleep_amount));
+			}
+			if (retry_cb) {
+				retry_cb();
+			}
+		} else {
+			throw IOException("HTTP " + method + " error on '" + url + "' (Error code " + to_string((int)res.error()) +
```

Comment:
If you remove the int cast here, we'll get the actual error code name instead (if the `to_string` overload is in scope)

Reply:
good point, thats much nicer!